### PR TITLE
Provide default nginx config and per-drupal-version config

### PR DESCRIPTION
**Pull request by rfay**
## The Problem:

In drud/ddev#648 (redirect module causes infinite redirects in drupal8) we discovered that we're trying to use a one-size-fits-all nginx config, when we really should be more flexible. 

We have been expecting to need nginx-config-per-project-type for some time, this PR introduces nginx config for each project type, with a default (which is the same as what we were using before). 

## The Manual Test:

This image is pushed as drud/nginx-php-fpm-local:20180224_multi_cms_nginx - use it in your .ddev/config.yaml in various project types, especially drupal8 and TYPO3

See https://github.com/drud/ddev/pull/669 for more testing instructions.

## Related Issue Link(s):

OP: drud/ddev#648 - where the d8 redirect module causes enormous misbehavior.
https://github.com/drud/ddev/pull/669 is the ddev PR to pull this in as a container.

## Release/Deployment notes:

After all PRs to go in a release have been pulled, a new release has to be created, docker image pushed, and ddev version updated with a PR.

### CODE REVIEW

**Comment by rfay** [comment]
I'll try to get some tests added and ping for your approval, but thanks for the approval. Good point about needing tests every time for new logic like this.

**Comment by rfay** [comment]
OK @cweagans 

* Added test coverage for backdrop additions
* Added test coverage for per-cms nginx config
* Bumped the timeouts in nginx and php
* Made a specific TYPO3 config that is not just drupalish.

Thanks for another quick look.

File: test/containertest.sh [hunk 1]
```diff
@@ -63,6 +63,14 @@ for project_type in drupal6 drupal7 drupal8 typo3 default; do
 	# Make sure that the right PHP version was selected for the project_type
 	# Only drupal6 is currently different here.
 	docker exec -it $CONTAINER php --version | grep "PHP $PHP_VERSION"
+
+	# Make sure that backdrop drush commands were added on backdrop and only backdrop
+	if [ "$project_type" == "backdrop" ] ; then
+	 	# fails if the directory exists
```
Comment (cweagans) [inline test/containertest.sh]: This should fail if the directory does _not_ exist, right?

Comment (rfay) [reply test/containertest.sh]: Yeah, I'll fix the comment and make the two tests match each other.

**Comment by cweagans** [comment]
:+1: lgtm

**Comment by rickmanelius** [comment]
Comparing each new nginx config to nginx-site-default.conf because it's easier to know exactly what sections to be mindful of for a smoke test.

Backdrop

```
➜  nginx git:(d20ee2f) diff nginx-site-backdrop.conf nginx-site-default.conf
1c1
< # ddev backdrop config
---
> # ddev default config
26c26
<         try_files $uri /index.php?$query_string; # For Drupal >= 7
---
>         try_files $uri $uri/ /index.php?q=$uri&$args;
105a106
>
```

Drupal 6


```
➜  nginx git:(d20ee2f) diff nginx-site-drupal6.conf nginx-site-default.conf
1c1
< # ddev drupal6 config
---
> # ddev default config
26c26
<         try_files $uri @rewrite; # For Drupal <= 6
---
>         try_files $uri $uri/ /index.php?q=$uri&$args;
29a30,31
>         # For D7 and above:
>         # Clean URLs are handled in drupal_environment_initialize().
33,34c35,36
<     # Handle imagecache
<     location ~* /imagecache/ {
---
>     # Handle image styles for Drupal 7+
>     location ~ ^/sites/.*/files/styles/ {
103a106
>
```

Drupal 7

```
➜  nginx git:(d20ee2f) diff nginx-site-drupal7.conf nginx-site-default.conf
1c1
< # ddev drupal7 config
---
> # ddev default config
26,27d25
<         # First attempt to serve request as file, then
<         # as directory, then fall back to index.html
107a106
>
```

Drupal 8

```
➜  nginx git:(d20ee2f) diff nginx-site-drupal8.conf nginx-site-default.conf
1c1
< # ddev drupal8 config
---
> # ddev default config
26c26
<         try_files $uri /index.php?$query_string; # For Drupal >= 7
---
>         try_files $uri $uri/ /index.php?q=$uri&$args;
105a106
>
```

TYPO3

```
➜  nginx git:(d20ee2f) diff nginx-site-typo3.conf nginx-site-default.conf
1c1
< # ddev typo3 config
---
> # ddev default config
26c26,37
<         try_files $uri $uri/ /index.php$is_args$args;
---
>         try_files $uri $uri/ /index.php?q=$uri&$args;
>     }
>
>     location @rewrite {
>         # For D7 and above:
>         # Clean URLs are handled in drupal_environment_initialize().
>         rewrite ^ /index.php;
>     }
>
>     # Handle image styles for Drupal 7+
>     location ~ ^/sites/.*/files/styles/ {
>         try_files $uri @rewrite;
71a83,93
>     ## Regular private file serving (i.e. handled by Drupal).
>     location ^~ /system/files/ {
>         ## For not signaling a 404 in the error log whenever the
>         ## system/files directory is accessed add the line below.
>         ## Note that the 404 is the intended behavior.
>         log_not_found off;
>         access_log off;
>         expires 30d;
>         try_files $uri @rewrite;
>     }
>
83,135d104
<
<     location = /favicon.ico {
<         log_not_found off;
<         access_log off;
<     }
<
<     # TYPO3 - Block access to composer files
<     location ~* composer\.(?:json|lock) {
<         deny all;
<     }
<
<     # TYPO3 - Block access to flexform files
<     location ~* flexform[^.]*\.xml {
<         deny all;
<     }
<
<     # TYPO3 - Block access to language files
<     location ~* locallang[^.]*\.(?:xml|xlf)$ {
<         deny all;
<     }
<
<     # TYPO3 - Block access to static typoscript files
<     location ~* ext_conf_template\.txt|ext_typoscript_constants\.(?:txt|typoscript)|ext_typoscript_setup\.(?:txt|typoscript) {
<         deny all;
<     }
<
<     # TYPO3 - Block access to miscellaneous protected files
<     location ~* /.*\.(?:bak|co?nf|cfg|ya?ml|ts|typoscript|dist|fla|in[ci]|log|sh|sql)$ {
<         deny all;
<     }
<
<     # TYPO3 - Block access to recycler and temporary directories
<     location ~ _(?:recycler|temp)_/ {
<         deny all;
<     }
<
<     # TYPO3 - Block access to configuration files stored in fileadmin
<     location ~ fileadmin/(?:templates)/.*\.(?:txt|ts|typoscript)$ {
<         deny all;
<     }
<
<     # TYPO3 - Block access to libaries, source and temporary compiled data
<     location ~ ^(?:vendor|typo3_src|typo3temp/var) {
<         deny all;
<     }
<
<     # TYPO3 - Block access to protected extension directories
<     location ~ (?:typo3conf/ext|typo3/sysext|typo3/ext)/[^/]+/(?:Configuration|Resources/Private|Tests?|Documentation|docs?)/ {
<         deny all;
<     }
<
<
<
136a106

```

**Comment by rfay** [comment]
Testing showed that imagecache (and other things) didn't work in drupal6, I pushed https://github.com/drud/docker.nginx-php-fpm-local/pull/51/commits/e5ad8d3393371930c3100f29282ff4def02d0c7e with a much better d6 config.

**Comment by rfay** [comment]
Added custom wordpress config, solved quite a number of problems with drupal6 and backdrop.

**Pull request merged**